### ISSUE TITLE
BrowserWindow.addDevToolsExtension, web_accessible_resources not supporting `chrome-extension://` protocol

### ISSUE THREAD

**Issue by codekiln** [electron/electron#8647]
* Electron version: 1.4.15
* Operating system: OS X 10.9.5

### Expected behavior

A chrome extension that has asset paths in the `web_accessible_resources` key of its `manifest.json` should be accessible via [`chrome-extension://[PACKAGE ID]/[PATH]`](https://developer.chrome.com/extensions/manifest/web_accessible_resources) after the extension has been correctly installed via [`BrowserWindow.addDevToolsExtension()`](http://electron.atom.io/docs/all/#browserwindowadddevtoolsextensionpath). After loading the chrome extension, loading a path to one of its web accessible resources in the browser should show the web accessible resource.

### Actual behavior

```
let installed = BrowserWindow.getDevToolsExtensions();
console.log("Installed chrome extensions:");
console.log(installed); // result - 'Cisco Web Communicator' is successfully installed
-> Installed chrome extensions:
-> { devtron: { name: 'devtron', version: '1.0' },
->   'React Developer Tools': { name: 'React Developer Tools', version: '0.15.6' },
->   'Redux DevTools': { name: 'Redux DevTools', version: '2.13.1' },
->   'Cisco Web Communicator': { name: 'Cisco Web Communicator', version: '3.1.0.363' } }
let extensionID = "ppbllmlcmhfnfflbkbinnhacecaankdh";
let urlToLoad = 'chrome-extension://' + extensionID + '/cwic_plugin.js';
mainWindow.loadURL(urlToLoad);
-> [64149:0209/144339:ERROR:CONSOLE(10146)] "Skipping extension with invalid URL: chrome-extension://cisco-web-communicator", source: chrome-devtools://devtools/bundled/inspector.js (10146)
-> chrome-extension://ppbllmlcmhfnfflbkbinnhacecaankdh/cwic_plugin.js Failed to load resource: net::ERR_NOT_IMPLEMENTED
```

### How to reproduce

```js
const installExtensions = async () => {
  const installer = require('electron-devtools-installer'); 
  const reqChromeExtensions = ['ppbllmlcmhfnfflbkbinnhacecaankdh'];
  const reqChromeExtenPromises = reqChromeExtensions.map(id => installExtension(id));
  return Promise.all(reqChromeExtenPromises).catch(console.log);
};

app.on('ready', async () => {
  await installExtensions();

  mainWindow = new BrowserWindow({
    show: false,
    width: 1024,
    height: 1600,
    webPreferences: {
      webSecurity: false
    }
  });

  let installed = BrowserWindow.getDevToolsExtensions();
  console.log("Installed chrome extensions:");
  console.log(installed);

  let extensionID = "ppbllmlcmhfnfflbkbinnhacecaankdh";
  let urlToLoad = 'chrome-extension://' + extensionID + '/cwic_plugin.js';
  mainWindow.loadURL(urlToLoad);
```

### Context
I'm attempting to use the [Cisco Jabber SDK](https://developer.cisco.com/site/jabber-websdk/develop-and-test/voice-and-video/downloads-and-docs/) inside an electron app. This requires installing a piece of software on the user's computer, installing a chrome extension, and including a Cisco script `cwic.js`, which extends the Jabber softphone API to the js context. `cwic.js` injects `chrome-extension://ppbllmlcmhfnfflbkbinnhacecaankdh/cwic_plugin.js` into the browser. I've already checked that all of the `chrome.*` apis referenced inside the chrome extension are supported by electron as mentioned in #5842 . I've read through https://github.com/electron/electron/blob/v1.4.15/lib/browser/chrome-extension.js#L73, and I can see that it doesn't look like `web_accessible_resources` is extracted from the `manifest.json` when extensions are loaded. Is this the case? If so, are there any known workarounds?

**Comment by KagamiChan** [comment]
I've done some search on this topic before, here's some findings that might help
- Electron does not support all Chrome extensions, the initial intention was to support some developer extensions. https://github.com/electron/electron/blob/master/docs/tutorial/devtools-extension.md
- `chrome-extension` protocol is currently available for devtool tabs.
- Not all chrome.* API are implemented, and they are only available for pages from `chrome-extension` protocol. https://github.com/electron/electron/blob/master/lib/renderer/init.js#L78

Please correct me if I went wrong.

I think you may implement a `chrome-extension` custom protocol similar to that in `chrome-extension.js`, and make up the missing chrome.* API if necessary. I'm not sure if there's some ipc problems.

**Comment by codekiln** [comment]
Hi @KagamiChan, thank you for your response. As I alluded to in my post, I'm aware that not all of the `chrome.*` apis are supported, and that's not the concern here because I've checked that all of the `chrome.*` apis I need are already implemented in Electron:

> I've already checked that all of the chrome.* apis referenced inside the chrome extension are supported by electron as mentioned in #5842 . I've read through https://github.com/electron/electron/blob/v1.4.15/lib/browser/chrome-extension.js#L73, and I can see that it doesn't look like web_accessible_resources is extracted from the manifest.json when extensions are loaded.

My goal here is to find out why the `chrome-extension://` protocol seems to reference extensions by name instead of id, and determine if the electron team is amenable to an enhancement bringing support for `web_accessible_resources`. If so, I can submit a PR to implement the `chrome-extension://[PACKAGE ID]/[PATH]` protocol to access items referenced in the `web_accessible_resources` of extensions' `manifest.json`.

**Comment by MarshallOfSound** [comment]
> determine if the electron team is amenable to an enhancement bringing support for web_accessible_resources. 

If it's part of the spec, not too complex (doesn't look like it) and adds support for more chrome extensions I'd personally love a PR for it :D

It sounds like something that we have to be super careful about though (potentially exposing filesystem to remote pages).

/cc @electron/maintainers Is this something that fits into the scope of chrome extension compatibility we want to provide?

**Comment by deepak1556** [comment]
I don't see the need to support parsing `web_accessible_resources` with our current implementation of `chrome-extension` protocol. Unlike chrome we haven't implemented a strict policy of what resources can be requested from the extension, all resources are responded with CORS headers. In your example if you try to load `chrome.runtime.getURL(<relative path>)` , it should work.

**Comment by codekiln** [comment]
@deepak1556, the need is not there if you are writing your own js code; in that case, one could use the `runtime.getURL` method. The need arises only if you want use a *third party* js library that already makes use of the `chrome-extension://` protocol to refer to a *third party* chrome extension. In this case, I would like to make a Cisco Jabber softphone application. For that I need to use Cisco's cwic.js library, which already calls `chrome-extension://` in a way that refers to Cisco's chrome extension.

**Comment by deepak1556** [comment]
> For that I need to use Cisco's cwic.js library, which already calls chrome-extension:// in a way that refers to web accessible resources in Cisco's chrome extension.

@codekiln I believe cwic.js uses a hardcoded extension id to request some existing resource, then yes it will be problem because the id will defer from what we have internally in Electron when installing the extension. I would say the problem is within the library itself not Electron.

On a side note, I am not familiar with jabber sdk but isn't there other versions of using the sdk without relying on chrome extension.

**Comment by codekiln** [comment]
@deepak1556 Chrome [uses and supports the standard of a hardcoded `chrome-extension://[PACKAGE ID]/[PATH]`](https://developer.chrome.com/extensions/manifest/web_accessible_resources), and Jabber uses that standard. [Electron even uses this standard when loading the background pages for each extension](https://github.com/electron/electron/blob/v1.4.15/lib/browser/chrome-extension.js#L98) and in the implementation of [`chrome.runtime.getURL` itself](https://github.com/electron/electron/blob/3834f0c6a6909194184d1f2ffbe457ebfa5c3d76/lib/renderer/chrome-api.js#L89). It's just that `web_accessible_resources` is currently unimplemented. 

With regards to your side note, Jabber SDK [requires the chrome extension since Chrome removed the NPAPI it formerly depended on a few years ago](https://communities.cisco.com/community/developer/jabber/blog/2014/12/11/jabber-guest-105--chrome-npapi-plugin-update-ios-mobile-hd).

**Comment by deepak1556** [comment]
@codekiln in your example you are trying to load `chrome-extension://ppbllmlcmhfnfflbkbinnhacecaankdh/cwic_plugin.js` where the hostname is the extension id but its not ID under which it is registered in Electron, rather it is the manifest name field, in this extension case try loading `chrome-extension://cisco-web-communicator/cwic_plugin.js`

**Comment by codekiln** [comment]
@deepak1556 why does electron [use the chrome extension's name as the extensionId](https://github.com/electron/electron/blob/master/lib/browser/chrome-extension.js#L50) instead of just using the extension's actual ID?

**Comment by deepak1556** [comment]
> why does electron use the chrome extension's name as the extensionId instead of just using the extension's actual ID? Is there a technical reason?

I dont think so, but will defer to @zcbenz since he implemented it.

**Comment by zcbenz** [comment]
> why does electron use the chrome extension's name as the extensionId instead of just using the extension's actual ID? Is there a technical reason?

The API may load from the extension's source code directly, which does not have an ID available. We should also support using real extension ID as hostname when there is one.

**Comment by codekiln** [comment]
It may be possible to compute the extension id from the crx bundle; see this
code
<https://github.com/shyiko/node-chrome-extension-id/blob/master/src/index.js>
from the chrome-extension-id
<https://www.npmjs.com/package/chrome-extension-id> npm package.

**Comment by RickyV33** [comment]
I'm currently facing the same issue. Was there any resolution to this problem?

**Comment by codekiln** [comment]
No. I offered to make a fix, but it wasn't clear that would be accepted, so
I didn't create the fixing PR. I ended up going to nw.js, which has better
support for chrome extensions.

**Comment by sofianguy** [comment]
Thanks for contributing!

We are no longer releasing bug fixes for versions of Electron older than 1.5. We prefer to only keep issues that are on our immediate radar.

If you are still having the same issue in an up-to-date version, feel free to reach out and we'll reopen.
